auth/src/main/java/com/firebase/ui/auth/ui/idp/AuthMethodPickerActivity.java

```diff
@@ -154,17 +155,28 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
     public void onSuccess(final IDPResponse response) {
         AuthCredential credential = createCredential(response);
         final FirebaseAuth firebaseAuth = mActivityHelper.getFirebaseAuth();
-
-        firebaseAuth
-                .signInWithCredential(credential)
-                .addOnFailureListener(
-                        new TaskFailureLogger(TAG, "Firebase sign in with credential unsuccessful"))
-                .addOnCompleteListener(new CredentialSignInHandler(
-                        AuthMethodPickerActivity.this,
-                        mActivityHelper,
-                        RC_ACCOUNT_LINK,
-                        RC_SAVE_CREDENTIAL,
-                        response));
+        FirebaseUser user = firebaseAuth.getCurrentUser();
+
+        TaskFailureLogger logger = new TaskFailureLogger(
+                TAG, "Firebase sign in with credential unsuccessful");
+        CredentialSignInHandler handler = new CredentialSignInHandler(
+                AuthMethodPickerActivity.this,
+                mActivityHelper,
+                RC_ACCOUNT_LINK,
+                RC_SAVE_CREDENTIAL,
+                response);
+
+        if (user != null && mActivityHelper.getFlowParams().shouldLinkUser) {
+            user
+                    .linkWithCredential(credential)
+                    .addOnFailureListener(logger)
+                    .addOnCompleteListener(handler);
+        } else {
+            firebaseAuth
+                    .signInWithCredential(credential)
```

Comment:
Seems like in this `if/else` block you are attaching the same complete/failure listener but switching on the Task type.  Would be clearer to do:

```java
// Either sign the user in or link the credential
Task<AuthResult> authTask;
if (user != null && mActivityHelper.getFlowParams().shouldLinkUser) {
  authTask = user.linkWithCredential(credential);
} else {
  authTask = firebaseAuth.signInWithCredential(credential);
}

task.addOnFailureListener(logger);
task.addOnCompleteListener(handler);
```